src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.SslCtx.cs

```diff
@@ -69,7 +83,107 @@ protected override bool ReleaseHandle()
         {
             Interop.Ssl.SslCtxDestroy(handle);
             SetHandle(IntPtr.Zero);
+            if (_gch.IsAllocated)
+            {
+                //Interop.Ssl.SslCtxSetData(this, (IntPtr)_gch);
+                _gch.Free();
+            }
+
+            if (_sslSessions != null)
+            {
+                lock (_sslSessions)
+                {
+                    foreach (string name in _sslSessions.Keys)
+                    {
+                        _sslSessions.Remove(name, out IntPtr session);
+                        Interop.Ssl.SessionFree(session);
+                    }
+                }
+            }
+
             return true;
         }
+
+        public void EnableSessionCache()
+        {
+            _sslSessions = new ConcurrentDictionary<string, IntPtr>();
+            _gch = GCHandle.Alloc(this);
+            // This is needed so we can find the handle from session remove callback.
+            Interop.Ssl.SslCtxSetData(this, (IntPtr)_gch);
+        }
+
+        public bool TryAddSession(IntPtr serverName, IntPtr session)
+        {
+            Debug.Assert(_sslSessions != null && session != IntPtr.Zero);
+
+            if (_sslSessions == null || serverName == IntPtr.Zero)
+            {
+                return false;
+            }
+
+            string? name = Marshal.PtrToStringAnsi(serverName);
+            if (!string.IsNullOrEmpty(name))
+            {
+                Interop.Ssl.SessionSetHostname(session, serverName);
+
+                lock (_sslSessions)
+                {
+                    IntPtr oldSession = _sslSessions.GetOrAdd(name, session);
+                    if (oldSession != session)
+                    {
+                        _sslSessions.Remove(name, out oldSession);
+                         Interop.Ssl.SessionFree(oldSession);
+                        oldSession = _sslSessions.GetOrAdd(name, session);
+                        Debug.Assert(oldSession == session);
+                    }
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        public void Remove(string name, IntPtr session)
+        {
+            if (_sslSessions != null)
+            {
+                lock (_sslSessions)
+                {
+                    if (!_sslSessions.Remove(name, out IntPtr oldSession))
+                    {
+                        Interop.Ssl.SessionFree(oldSession);
+                    }
+                }
+            }
+        }
+
+        public bool TrySetSession(SafeSslHandle sslHandle, string name)
+        {
+            Debug.Assert(_sslSessions != null);
+
+            if (_sslSessions == null || string.IsNullOrEmpty(name))
+            {
+                return false;
+            }
+
+            // even if we don't have matching session, we can get new one and we need
+            // way how to link SSL back to `this`.
+            Interop.Ssl.SslSetData(sslHandle, (IntPtr)_gch);
```

Comment:
In other parts of the codebase, we use `SslSetData` to list of ALPN protocols on the server side, can't it get overwritten then?

Reply:
correct. We we set would differ between client and server. We would have or the other but we should not have both at the same time. 
OpenSSL allows to set more then one but then there is registration process and I would like to avoid it as setting one is cheap and simple

Reply:
I added some Assert to make sure it is not set. This would be easy to miss in future changes.